### PR TITLE
Threaded datum parsing

### PR DESCRIPTION
**Pull request by slayton58**
Move I/O datum parsing from the main read thread into the decoder / transform threads

### CODE REVIEW

**Comment by drnikolaev** [comment]
@slayton58 I wonder what we gain by using `string*` instead of just `string`? Avoiding copying them? Something else?

**Comment by slayton58** [comment]
@drnikolaev 1) I shadowed the behaviour of the original Datum and 2) the string containing the pre-parsed Datum is O(size of the datum), so we definitely don't want to be copying it around (imagine the string contains full uncompressed images in the worst case - we only want a single copy of that data).

**Comment by drnikolaev** [comment]
@slayton58 Thanks. My concern is about using naked pointer here. I'd suggest to wrap it by `shared_ptr`. In this case `~QueuePair` would look much cleaner.

**Comment by slayton58** [comment]
@drnikolaev The issue here would be any subsequent merging into BVLC - The entirety of the Datum / string passing in this section is done via raw pointers, so we'd either have to rip it out again or persuade them to move to shared_ptr in this instance.

My feeling here is that the code is isolated enough, has little complexity and is easy enough to manually manage that I don't see raw pointers as a problem (but then I don't ever really see raw pointers as a problem, so ymmv).

**Comment by drnikolaev** [comment]
@slayton58 Ok, thanks.

File: src/caffe/data_reader.cpp [hunk 1]
```diff
@@ -103,9 +103,9 @@ void DataReader::Body::InternalThreadEntry() {
 }
 
 void DataReader::Body::read_one(db::Cursor* cursor, QueuePair* qp) {
-  Datum* datum = qp->free_.pop();
+  string* datum = qp->free_.pop();
   // TODO deserialize in-place instead of copy?
```
Comment (gheinrich) [inline src/caffe/data_reader.cpp:107]: is this comment still relevant?

Comment (slayton58) [inline src/caffe/data_reader.cpp:107]: Not really, actually.

File: src/caffe/layers/data_layer.cpp [hunk 2]
```diff
@@ -112,11 +118,12 @@ void DataLayer<Dtype>::load_batch(Batch<Dtype>* batch) {
   for (int item_id = 0; item_id < batch_size; ++item_id) {
       timer.Start();
       // Get a datum
-      Datum* datum = (reader_.full().pop("Waiting for data"));
+      string* str = (reader_.full().pop("Waiting for data"));
```
Comment (gheinrich) [inline src/caffe/layers/data_layer.cpp:121]: question: how do you handle the case where you can't load a full batch (when the batch size does not divide the number of items in the database)?

File: src/caffe/util/blocking_queue.cpp [hunk 3]
```diff
@@ -88,7 +88,8 @@ size_t BlockingQueue<T>::size() const {
 
 template class BlockingQueue<Batch<float>*>;
 template class BlockingQueue<Batch<double>*>;
-template class BlockingQueue<Datum*>;
+// template class BlockingQueue<Datum*>;
```
Comment (gheinrich) [inline src/caffe/util/blocking_queue.cpp:91]: this can be removed I think

File: src/caffe/layers/data_layer.cpp (hunk 2 above)
Comment (slayton58) [inline src/caffe/layers/data_layer.cpp:121]: You fail out (as the parallel code always has): https://github.com/NVIDIA/caffe/blob/caffe-0.14/src/caffe/parallel.cpp#L454

File: src/caffe/util/blocking_queue.cpp (hunk 3 above)
Comment (slayton58) [inline src/caffe/util/blocking_queue.cpp:91]: Agreed, I forgot to remove it completely after checking it wasn't needed for compilation

File: src/caffe/layers/data_layer.cpp (hunk 2 above)
Comment (gheinrich) [inline src/caffe/layers/data_layer.cpp:121]: Isn't it a different thing? I'm asking how you handle the case where e.g. your batch size is 32 and you have 1000 entries in your dataset. I'm actually trying to understand this in the context of a tiny error I noticed in the calculation of accuracy.

Comment (slayton58) [inline src/caffe/layers/data_layer.cpp:121]: That's handled here: https://github.com/slayton58/caffe/blob/thread_pool_parsing/src/caffe/data_reader.cpp#L112-L116

If you try to run past the end of the DB, you loop around and start again from the beginning

Comment (gheinrich) [inline src/caffe/layers/data_layer.cpp:121]: Thanks!

**Comment by drnikolaev** [comment]
Are we still on hold with this one?

**Comment by slayton58** [comment]
I think it just needs someone else to test, then merge.

I haven't seen questions or concerns in a few days now

**Comment by drnikolaev** [comment]
Merged to let it go to QA on Monday morning.

**Pull request merged**